src/python/nxt_python.c

```diff
@@ -428,27 +429,42 @@ nxt_python_set_target(nxt_task_t *task, nxt_python_target_t *target,
         nxt_python_print_exception();
         goto fail;
     }
+    
+    char factory_found = 0;
+    value = nxt_conf_get_object_member(conf, &factory_str, NULL);
 
-    value = nxt_conf_get_object_member(conf, &callable_str, NULL);
-    if (value == NULL) {
-        callable = nxt_alloca(12);
-        nxt_memcpy(callable, "application", 12);
-
-    } else {
+    if (value != NULL) {
+        factory_found = 1;
         nxt_conf_get_string(value, &str);
-
-        callable = nxt_alloca(str.length + 1);
-        nxt_memcpy(callable, str.start, str.length);
-        callable[str.length] = '\0';
+        factory_name = nxt_alloca(str.length + 1);
+        nxt_memcpy(factory_name, str.start, str.length);
+        factory_name[str.length] = '\0';
+        nxt_alert(task, "factory name %s", factory_name);
+        obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
+        obj = PyObject_CallObject(obj, NULL);
     }
-
-    obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
     if (nxt_slow_path(obj == NULL)) {
         nxt_alert(task, "Python failed to get \"%s\" from module \"%s\"",
                   callable, module_name);
         goto fail;
     }
 
+    if (!factory_found) {
+        value = nxt_conf_get_object_member(conf, &callable_str, NULL);
+        if (value == NULL) {
+            callable = nxt_alloca(12);
+            nxt_memcpy(callable, "application", 12);
+
+        } else {
+            nxt_conf_get_string(value, &str);
+
+            callable = nxt_alloca(str.length + 1);
+            nxt_memcpy(callable, str.start, str.length);
+            callable[str.length] = '\0';
+        }
+        obj = PyDict_GetItemString(PyModule_GetDict(module), callable);
+    }
+
```

Comment:
I have a feeling all this logic could be cleaned up and possibly removing the need for the factory_found boolean...